### DISCUSSION TITLE
NativeAOT compatibility

### DISCUSSION THREAD

**Discussion by barefists** [yantrajs/yantra#102] [General]
I'm investigating using YantraJS to port some javascript code to an environment that can only run NativeAOT binaries.
I see that NativeAOT compatibility is in the roadmap for YantraJS, and I'm wondering if you have an estimate for when that might be possible?

**Comment by ackava** [comment]
We have already completed migration of Reflection to Generated classes, so the issue is partially fixed. But we are not sure if that will fix all issues, NativeAOT is really very big topic and we will try to fix the issues reported.

So if you try to run the current YantraJS and let me know if you face any issues, please report and I will let you know if those can be fixed easily or not. Mostly we spend time on YantraJS on weekends as currently there are no sponsors.

**Comment by barefists** [comment]
Thanks for your response. I'm doing some simple tests and getting an error on Eval:
`Exception thrown: 'System.PlatformNotSupportedException' in System.Private.CoreLib.dll
An unhandled exception of type 'System.PlatformNotSupportedException' occurred in System.Private.CoreLib.dll
Dynamic code generation is not supported on this platform.`

Is there a way of compiling the JS ahead of time?

**Reply by ackava** [reply]
Since YantraJS dynamically generates IL, and it is not possible to precompile as things like Eval and Function constructor need to dynamically generate the code. And the way JavaScript works, full AOT will not work. 

However under AOT you can enable Mono Interpreter that will allow YantraJS to work correctly.

We have successfully executed AOT with interpreter on iOS.

**Comment by barefists** [comment]
Hi, I'm starting a new project from scratch (the previous one couldn't be ported over to YantraJS because a lot of reflection was used)
I'm looking for NativeAOT-friendly ways of calling JS from CSharp using Yantra.
At the moment, Execute, Eval and RunAsync all break the NativeAOT workflow due to runtime parsing and/or reflection use.
Am I missing something?

**Comment by ackava** [comment]
DotNet offers MSIL interpretation in NativeAOT workflow (You have to enable Mono Interpreter). YantraJS's initialization sequence was using lot of reflection that has been removed and is now using minimal Reflection.

If you are facing any difficulty, then please create a sample repo that I can see the issue and will try to find workaround.
